File: DataProcess/Astar.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <cmath>
#include <algorithm>

using namespace std;

class Node {
public:
    pair<int, int> position;
    Node* parent;
    int g;
    int h;
    int f;

    Node(pair<int, int> pos, Node* par = nullptr) : position(pos), parent(par), g(0), h(0), f(0) {
        f = g + h;
    }

    bool operator<(const Node& other) const {
        return f < other.f;
    }

    bool operator==(const Node& other) const {
        return position == other.position;
    }
};

double get_distance(pair<int, int> pos1, pair<int, int> pos2) {
    return sqrt(pow(pos1.first - pos2.first, 2) + pow(pos1.second - pos2.second, 2));
}

vector<pair<int, int>> Astar(pair<int, int> start, pair<int, int> goal, vector<vector<int>>& grid) {
    vector<Node*> open_list;
    vector<Node*> closed_list;
    Node* start_node = new Node(start);
    Node* goal_node = new Node(goal);
    open_list.push_back(start_node);
    vector<pair<int, int>> directions = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};

    while (!open_list.empty()) {
        Node* current_node = *min_element(open_list.begin(), open_list.end(), [](const Node* a, const Node* b) {
            return a->f < b->f;
        });
        open_list.erase(remove(open_list.begin(), open_list.end(), current_node), open_list.end());
        closed_list.push_back(current_node);

        if (current_node->position == goal_node->position) {
            vector<pair<int, int>> path;
            while (current_node->parent) {
                path.push_back(current_node->position);
                current_node = current_node->parent;
            }
            path.push_back(start_node->position);
            reverse(path.begin(), path.end());
            return path;
        }

        for (auto direction : directions) {
            int next_x = current_node->position.first + direction.first;
            int next_y = current_node->position.second + direction.second;
            Node* next_node = new Node({next_x, next_y}, current_node);

            if (next_x < 0 || next_x >= grid.size() || next_y < 0 || next_y >= grid[0].size() || grid[next_x][next_y] == 1) {
                continue;
            }

            next_node->g = current_node->g + get_distance(next_node->position, current_node->position);
            next_node->h = abs(next_x - goal_node->position.first) + abs(next_y - goal_node->position.second);
            next_node->f = next_node->g + next_node->h;

            auto it_closed = find_if(closed_list.begin(), closed_list.end(), [&](Node* n) { return *n == *next_node; });
            if (it_closed != closed_list.end()) {
                continue;
            }

            auto it_open = find_if(open_list.begin(), open_list.end(), [&](Node* n) { return *n == *next_node; });
            if (it_open != open_list.end()) {
                Node* existing_node = *it_open;
                if (existing_node->f > next_node->f) {
                    open_list.erase(remove(open_list.begin(), open_list.end(), existing_node), open_list.end());
                }
            }

            open_list.push_back(next_node);
        }
    }

    return vector<pair<int, int>>();
}

ostream& operator<<(ostream& cout , const pair<int,int>& point){
    cout << "(" << point.first << "," << point.second << ")" << " -> ";
    return cout;
}

// vector<vector<int>> read_map(string filename) {
//        //
// }
vector<vector<int>> map = {

};

int main() {
    vector<vector<int>> map = {
        {0, 0, 0, 0, 0, 0},
        {0, 0, 0, 0, 0, 0},
        {0, 0, 0, 0, 0, 0},
        {0, 0, 0, 0, 0, 0},
        {0, 0, 0, 1, 1, 0}
    };
    pair<int, int> start = make_pair(0, 0);
    pair<int, int> goal = make_pair(2, 3);
    vector<pair<int, int>> res = Astar(start, goal, map);
    for(pair<int,int> p : res){
        cout << p;
    }

    return 0;
}

File: DataProcess/Astar.py
'''
根据灰度图获得最佳的路径
'''
import math
import heapq
import time
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt 
from scipy.interpolate import interp1d

class Node:
    def __init__(self, position, parent=None):
        self.position = position
        self.parent = parent
        self.g = 0
        self.h = 0
        self.f = 0
        self.f = self.g + self.h

    def __lt__(self, other):
        return self.f < other.f

    def __eq__(self, other):
        return self.position == other.position

def gdist(position1, position2):
    dist : float = math.sqrt((position1[0] - position2[0]) ** 2 + (position1[1] - position2[1]) ** 2)
    return dist

def astar(start, goal, grid):
    open_list = []
    closed_list = []
    start_node = Node(start)
    goal_node = Node(goal)
    heapq.heappush(open_list, start_node)
    # directions = [(1, 0), (-1, 0), (0, 1), (0, -1)]
                #   ,(-1, -1), (-1, 1), (1, -1), (1, 1)]
    directions = [(1, 0), (-1, 0), (0, 1), (0, -1),(-1,-1),(-1,1),(1,-1),(1,1)]

    while open_list:
        current_node = heapq.heappop(open_list)
        closed_list.append(current_node)
        
        if current_node.position == goal_node.position:
            path = []
            while current_node.parent:
                path.append(current_node.position)
                current_node = current_node.parent
            path.append(start_node.position)
            return path[::-1]

        for direction in directions:
            next_x = current_node.position[0] + direction[0]
            next_y = current_node.position[1] + direction[1]
            next_node = Node((next_x, next_y), current_node)

            if next_x < 0 or next_x >= len(grid) or next_y < 0 or next_y >= len(grid[0]) or grid[next_x][next_y] == 0:
                continue

            next_node.g = current_node.g + gdist(next_node.position, current_node.position)
            next_node.h = abs(next_x - goal_node.position[0]) + abs(next_y - goal_node.position[1])
            next_node.f = next_node.g + next_node.h

            if next_node in closed_list:
                continue

            if next_node in open_list:
                for node in open_list:
                    if node == next_node and node.f > next_node.f:
                        open_list.remove(node)
                        break

            heapq.heappush(open_list, next_node)

    return None


data = pd.read_excel("../data/map.xlsx" , header = None)
data = data.to_numpy()
# plt.imshow(data , cmap = 'binary' , interpolation = 'nearest')
# plt.axis("off")
# plt.savefig("../data/binaryMap.png")
start = (282,377)
goal = (0,25)

path = astar(start, goal, data)
print(path)

x = [point[0] for point in path]
y = [len(data[0]) - point[1] for point in path]
# f = interp1d(x, y, kind='linear')
# x_interp = np.linspace(min(x), max(x), 100)
# y_interp = f(x_interp)
# plt.plot(x_interp, y_interp)
# plt.show()

plt.plot(y,x)
plt.gca().invert_yaxis()
plt.gca().invert_xaxis()
plt.gca().xaxis.set_ticklabels([])
plt.gca().yaxis.set_ticklabels([])
plt.scatter(len(data[0]) - start[1],start[0],color = "red" ,s = 200)
plt.scatter(len(data[0]) - goal[1],goal[0],color = "green" , s = 200)
# plt.show()
plt.savefig("../image/path.png")
with open("../data/pathpoint.txt","w") as file:
    for point in path:
        file.write("(" + str(point[0]) + " , " + str(point[1]) + ")" + "\n")
